### PR TITLE
Add new implementation of ParseGlobalArgs

### PR DESCRIPTION
**Pull request by nicktobey**
The main benefit of this new implementation is that it won't incorrectly parse situations where a global arg value (such as username) matches a dolt subcommand. (This would have been especially problematic when we got around to converting `dolt admin`.)

This also adds a nifty little burndown chart of commands that still need to be converted from using dEnv to using CliContext, in order to determine whether to generate the CliContext for a command. It'll be extra satisfying to remove elements from that list as we convert them.

### CODE REVIEW

File: go/libraries/utils/argparser/parser.go [hunk 1]
```diff
@@ -281,90 +308,103 @@ func (ap *ArgParser) Parse(args []string) (*ArgParseResults, error) {
...
+		return nil, ap.TooManyArgsErrorFunc(list)
+	}
 
-			results[opt.Name] = ""
-		}
+	return &ArgParseResults{results, list, ap}, nil
+}
 
-		opt, value := ap.matchValueOption(rest, isLongFormFlag)
+func (ap *ArgParser) parseToken(args []string, i int, list []string, results map[string]string) (int, []string, map[string]string, error) {
```
Comment (macneale4) [inline go/libraries/utils/argparser/parser.go:336]: Can you add names to the output variables.  It's not clear what is returned, and I suspect all you need it to add descriptive names - i -> index, list -> ???, results -> foundArguments.

This is one of these cases where gDoc would be reasonable, but probably not required if you had more descriptive names.

File: go/libraries/utils/argparser/parser.go [hunk 2]
```diff
@@ -271,6 +271,33 @@ func (ap *ArgParser) matchValueOption(arg string, isLongFormFlag bool) (match *O
...
+
+	i := 0
+	for ; i < len(args); i++ {
+		arg := args[i]
+
+		if len(arg) == 0 || arg == "--" {
+			continue
+		}
+
+		if arg[0] != '-' {
```
Comment (macneale4) [inline go/libraries/utils/argparser/parser.go:286]: Probably worth a one line comment here that no subcommand starts with '-', and that we assume this is a valid subcommand.

I don't see anyplace were we reject totally bogus commands - or having no command at all. Could remaining list be len(0) and line 531 in dolt.go result in a out of range error?

File: integration-tests/bats/no-repo.bats [hunk 3]
```diff
@@ -363,3 +363,10 @@ NOT_VALID_REPO_ERROR="The current directory is not a valid dolt repository."
     [ "$output" == "" ]
     run kill -9 $PID # Kill process if it doesn't pass
 }
+
+@test "no-repo: check that we correctly parse args when connecting with a username that matches a subcommand" {
```
Comment (macneale4) [inline integration-tests/bats/no-repo.bats:367]: please add a test (go or bats) which balks at bogus subcommands.

File: go/cmd/dolt/dolt.go [hunk 4]
```diff
@@ -120,6 +118,67 @@ var doltSubCommands = []cli.Command{
 	stashcmds.StashCommands,
 	&commands.Assist{},
 }
+
+var subCommandsUsingDEnv = []cli.Command{
```
Comment (macneale4) [inline go/cmd/dolt/dolt.go:122]: Should these be flags on a function or field in the cli.Command struct? Seems a little heavy weight.

Is your plan here to remove commands until there are no more?

File: go/libraries/utils/argparser/parser.go (hunk 1 above)
Comment (nicktobey) [reply go/libraries/utils/argparser/parser.go]: Done

File: go/cmd/dolt/dolt.go (hunk 4 above)
Comment (PavelSafronov) [reply go/cmd/dolt/dolt.go:122]: A flag or an empty interface would be a good approach here. Having a source of truth so far from the actual code can be problematic.

File: go/libraries/utils/argparser/parser.go (hunk 2 above)
Comment (nicktobey) [reply go/libraries/utils/argparser/parser.go:286]: If there's no subcommand we return an error on line 299.

`remaining` always contains at least the subcommand, so out of range is not possible there.

I added a comment.

File: integration-tests/bats/no-repo.bats (hunk 3 above)
Comment (nicktobey) [reply integration-tests/bats/no-repo.bats:367]: Done.

File: go/cmd/dolt/dolt.go [hunk 5]
```diff
@@ -470,7 +528,7 @@ The sql subcommand is currently the only command that uses these flags. All othe
 	}
 
 	var cliCtx cli.CliContext = nil
-	if initCliContext {
+	if initCliContext(remainingArgs[0]) {
```
Comment (PavelSafronov) [inline go/cmd/dolt/dolt.go:531]: Assign `remainingArgs[0]` to a named variable, so it's clear that this is the command name.

File: go/libraries/utils/argparser/parser.go (hunk 2 above)
Comment (macneale4) [reply go/libraries/utils/argparser/parser.go:286]: Huge improvement! Bravo!

**Review by macneale4 — APPROVED** [review]

File: go/cmd/dolt/dolt.go (hunk 5 above)
Comment (nicktobey) [reply go/cmd/dolt/dolt.go]: Done

**Pull request merged**